src/bridge/error/error.cc

```diff
@@ -0,0 +1,81 @@
+/**
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "src/bridge/error/error.h"
+
+#include <iostream>
+
+#include <openssl/err.h>
+
+#include "src/backing/status/status.h"
+#include "src/bridge/error/error_strings.h"
+
+namespace kmsengine {
+namespace bridge {
+
+// OpenSSL engines request a library code using ERR_get_next_error_library(3)
+// with which they can register engine-specific, human-readable error strings
+// with OpenSSL.
+static const int kErrorLibraryCode = ERR_get_next_error_library();
+
+Status LoadErrorStringsIntoOpenSSL() {
+  if (!ERR_load_strings(0, error_impl::kLibraryStrings) ||
+      !ERR_load_strings(GetLibraryCode(), error_impl::kReasonStrings)) {
+    return Status(StatusCode::kInternal, "ERR_load_strings failed");
+  }
+  return Status();
+}
+
+Status UnloadErrorStringsFromOpenSSL() {
+  if (!ERR_unload_strings(0, error_impl::kLibraryStrings) ||
+      !ERR_unload_strings(GetLibraryCode(), error_impl::kReasonStrings)) {
+    return Status(StatusCode::kInternal, "ERR_unload_strings failed");
+  }
+  return Status();
+}
+
+void SignalErrorToOpenSSL(Status status, const char *function_name,
+                          const char *file_name, int line_number) {
+  // `ERR_put_error` signals to OpenSSL that an error occurred. The second
+  // argument is a "function code"; we're ignoring that here by setting it to
+  // zero.
+  //
+  // This is similar to how OpenSSL 3.0.0 uses the `ERR_put_error` API - it
+  // has completely removed function codes in favor of automatically passing
+  // the function name from the `__func__` macro to `ERR_add_error_data`. The
+  // engine uses this approach since function names aren't very useful to the
+  // end-user anyways.
+  //
+  // Example from OpenSSL 3.0.0: https://github.com/openssl/openssl/pull/9072/
+  // files#diff-1d66b2010eedc54493f1641feb6dc5eaR375-R381
+  const auto kIgnoredFunctionCode = 0;
+  ERR_PUT_error(GetLibraryCode(), kIgnoredFunctionCode,
```

Comment:
The Google convention is to write `/*func=*/0` to explain literal arguments rather than defining a constant.